File: backend/src/main/java/com/edutech/jobportalsystem/service/TranslationService.java
package com.edutech.jobportalsystem.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class TranslationService {

    private static final Logger logger = LoggerFactory.getLogger(TranslationService.class);
    private final RestTemplate restTemplate = new RestTemplate();
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final Map<String, String> translationCache = new ConcurrentHashMap<>();

    @Value("${translation.enabled:true}")
    private boolean translationEnabled;

    @Value("${libretranslate.enabled:true}")
    private boolean libreTranslateEnabled;

    @Value("${libretranslate.url:https://api.libretranslate.com/translate}")
    private String libreTranslateUrl;

    @Value("${mymemory.enabled:true}")
    private boolean myMemoryEnabled;

    @Value("${mymemory.url:https://api.mymemory.translated.net/get}")
    private String myMemoryUrl;

    @Value("${google.translate.api-key:}")
    private String googleApiKey;

    @Value("${google.translate.enabled:false}")
    private boolean googleTranslateEnabled;

    @Value("${aws.translate.enabled:false}")
    private boolean awsTranslateEnabled;

    @Value("${deepl.api-key:}")
    private String deeplApiKey;

    @Value("${deepl.url:https://api-free.deepl.com/v1/translate}")
    private String deeplUrl;

    @Value("${deepl.enabled:false}")
    private boolean deeplEnabled;

    public String translate(String text, String targetLanguage) {
        if (!translationEnabled || text == null || text.isBlank()) {
            return text;
        }

        String cacheKey = targetLanguage + "::" + text;
        if (translationCache.containsKey(cacheKey)) {
            return translationCache.get(cacheKey);
        }

        String result = text;

        // Try each service in fallback order
        result = tryLibreTranslate(text, targetLanguage, result);
        if (!result.equals(text)) {
            translationCache.put(cacheKey, result);
            return result;
        }

        result = tryMyMemory(text, targetLanguage, result);
        if (!result.equals(text)) {
            translationCache.put(cacheKey, result);
            return result;
        }

        result = tryGoogleTranslate(text, targetLanguage, result);
        if (!result.equals(text)) {
            translationCache.put(cacheKey, result);
            return result;
        }

        result = tryAwsTranslate(text, targetLanguage, result);
        if (!result.equals(text)) {
            translationCache.put(cacheKey, result);
            return result;
        }

        result = tryDeepL(text, targetLanguage, result);
        translationCache.put(cacheKey, result);
        return result;
    }

    public String translateToEnglish(String text) {
        return translate(text, "EN");
    }

    private String tryLibreTranslate(String text, String targetLanguage, String fallback) {
        if (!libreTranslateEnabled) return fallback;
        try {
            String payload = "{\"q\":\"" + escapeJson(text) + "\",\"source\":\"auto\",\"target\":\"" + targetLanguage.toLowerCase() + "\"}";
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            HttpEntity<String> request = new HttpEntity<>(payload, headers);
            String response = restTemplate.postForObject(libreTranslateUrl, request, String.class);
            JsonNode node = objectMapper.readTree(response);
            String translated = node.get("translatedText").asText();
            if (translated != null && !translated.isBlank()) {
                logger.debug("LibreTranslate: Translated {} chars", text.length());
                return translated;
            }
        } catch (Exception e) {
            logger.debug("LibreTranslate failed: {}", e.getMessage());
        }
        return fallback;
    }

    private String detectLanguage(String text) {
        if (text == null || text.length() < 12) return "en";
        
        String[] germanWords = {"und", "der", "die", "das", "ein", "eine", "mit", "für", "auf"};
        String[] spanishWords = {"y", "de", "la", "el", "los", "las", "para", "con", "por"};
        String[] frenchWords = {"et", "le", "les", "des", "pour", "dans", "sur", "une"};
        
        String lowerText = text.toLowerCase();
        int germanScore = 0, spanishScore = 0, frenchScore = 0;
        
        for (String word : germanWords) {
            germanScore += countWord(lowerText, word);
        }
        for (String word : spanishWords) {
            spanishScore += countWord(lowerText, word);
        }
        for (String word : frenchWords) {
            frenchScore += countWord(lowerText, word);
        }
        
        // Check for non-Latin scripts
        if (text.matches(".*[\\u0400-\\u04FF].*")) return "ru"; // Cyrillic
        if (text.matches(".*[\\u3040-\\u30FF].*")) return "ja"; // Japanese
        if (text.matches(".*[\\u4E00-\\u9FFF].*")) return "zh"; // Chinese
        
        if (germanScore > spanishScore && germanScore > frenchScore && germanScore > 0) return "de";
        if (spanishScore > germanScore && spanishScore > frenchScore && spanishScore > 0) return "es";
        if (frenchScore > germanScore && frenchScore > spanishScore && frenchScore > 0) return "fr";
        
        return "en";
    }
    
    private int countWord(String text, String word) {
        int count = 0;
        int index = 0;
        while ((index = text.indexOf(word, index)) != -1) {
            // Check word boundaries
            if ((index == 0 || !Character.isLetterOrDigit(text.charAt(index - 1))) &&
                (index + word.length() == text.length() || !Character.isLetterOrDigit(text.charAt(index + word.length())))) {
                count++;
            }
            index += word.length();
        }
        return count;
    }

    private String tryMyMemory(String text, String targetLanguage, String fallback) {
        if (!myMemoryEnabled) return fallback;
        try {
            String sourceLang = detectLanguage(text);
            String targetLang = targetLanguage.equals("EN") ? "en" : targetLanguage.toLowerCase();
            
            // Skip if already in target language
            if (sourceLang.equals(targetLang)) return fallback;
            
            String url = myMemoryUrl + "?q=" + java.net.URLEncoder.encode(text, "UTF-8") + "&langpair=" + sourceLang + "|" + targetLang;
            String response = restTemplate.getForObject(url, String.class);
            JsonNode node = objectMapper.readTree(response);
            String translated = node.get("responseData").get("translatedText").asText();
            if (translated != null && !translated.isBlank() && !translated.equals(text)) {
                logger.debug("MyMemory: Translated {} chars from {} to {}", text.length(), sourceLang, targetLang);
                return translated;
            }
        } catch (Exception e) {
            logger.debug("MyMemory failed: {}", e.getMessage());
        }
        return fallback;
    }

    private String tryGoogleTranslate(String text, String targetLanguage, String fallback) {
        if (!googleTranslateEnabled || googleApiKey == null || googleApiKey.startsWith("YOUR_")) {
            return fallback;
        }
        try {
            String payload = "{\"q\":\"" + escapeJson(text) + "\",\"target\":\"" + targetLanguage.toLowerCase() + "\",\"key\":\"" + googleApiKey + "\"}";
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            HttpEntity<String> request = new HttpEntity<>(payload, headers);
            String response = restTemplate.postForObject(
                "https://translation.googleapis.com/language/translate/v2",
                request,
                String.class
            );
            JsonNode node = objectMapper.readTree(response);
            String translated = node.get("data").get("translations").get(0).get("translatedText").asText();
            if (translated != null && !translated.isBlank()) {
                logger.debug("Google Translate: Translated {} chars", text.length());
                return translated;
            }
        } catch (Exception e) {
            logger.debug("Google Translate failed: {}", e.getMessage());
        }
        return fallback;
    }

    private String tryAwsTranslate(String text, String targetLanguage, String fallback) {
        if (!awsTranslateEnabled) {
            return fallback;
        }
        // AWS requires SDK setup - placeholder for future implementation
        logger.debug("AWS Translate: Skipping (requires AWS SDK)");
        return fallback;
    }

    private String tryDeepL(String text, String targetLanguage, String fallback) {
        if (!deeplEnabled || deeplApiKey == null || deeplApiKey.startsWith("YOUR_")) {
            return fallback;
        }
        try {
            String payload = "{\"text\":\"" + escapeJson(text) + "\",\"target_lang\":\"" + targetLanguage + "\",\"auth_key\":\"" + deeplApiKey + "\"}";
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            HttpEntity<String> request = new HttpEntity<>(payload, headers);
            String response = restTemplate.postForObject(deeplUrl, request, String.class);
            JsonNode node = objectMapper.readTree(response);
            String translated = node.get("translations").get(0).get("text").asText();
            if (translated != null && !translated.isBlank()) {
                logger.debug("DeepL: Translated {} chars", text.length());
                return translated;
            }
        } catch (Exception e) {
            logger.debug("DeepL failed: {}", e.getMessage());
        }
        return fallback;
    }

    private String escapeJson(String text) {
        return text
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r");
    }

    private String encodeValue(String value) {
        try {
            return java.net.URLEncoder.encode(value, "UTF-8");
        } catch (java.io.UnsupportedEncodingException e) {
            return value;
        }
    }

    public void clearCache() {
        translationCache.clear();
        logger.info("Translation cache cleared");
    }

    public boolean isConfigured() {
        return translationEnabled && (libreTranslateEnabled || myMemoryEnabled);
    }
}

File: tools/scraper/translation-service.js
const axios = require('axios');

const translationCache = new Map();

// Simple language detection based on character patterns and word hints
function detectLanguage(text) {
  if (!text || text.length < 12) return 'en';
  
  const german_words = ['und', 'der', 'die', 'das', 'ein', 'eine', 'mit', 'für', 'auf'];
  const spanish_words = ['y', 'de', 'la', 'el', 'los', 'las', 'para', 'con', 'por'];
  const french_words = ['et', 'le', 'les', 'des', 'pour', 'dans', 'sur', 'une'];
  
  const lowerText = text.toLowerCase();
  let germanScore = 0, spanishScore = 0, frenchScore = 0;
  
  german_words.forEach(w => germanScore += (lowerText.match(new RegExp(`\\b${w}\\b`, 'g')) || []).length);
  spanish_words.forEach(w => spanishScore += (lowerText.match(new RegExp(`\\b${w}\\b`, 'g')) || []).length);
  french_words.forEach(w => frenchScore += (lowerText.match(new RegExp(`\\b${w}\\b`, 'g')) || []).length);
  
  // Check for non-Latin characters
  if (/[\u0400-\u04FF]/.test(text)) return 'ru'; // Cyrillic for Russian
  if (/[\u3040-\u30FF]/.test(text)) return 'ja'; // Japanese
  if (/[\u4E00-\u9FFF]/.test(text)) return 'zh'; // Chinese
  if (/[\uAC00-\uD7AF]/.test(text)) return 'ko'; // Korean
  
  if (germanScore > spanishScore && germanScore > frenchScore && germanScore > 0) return 'de';
  if (spanishScore > germanScore && spanishScore > frenchScore && spanishScore > 0) return 'es';
  if (frenchScore > germanScore && frenchScore > spanishScore && frenchScore > 0) return 'fr';
  
  return 'en';
}

class TranslationService {
  constructor() {
    this.enabled = process.env.TRANSLATION_ENABLED !== 'false';
    this.libreTranslateUrl = process.env.LIBRETRANSLATE_URL || 'https://api.libretranslate.com/translate';
    this.myMemoryUrl = process.env.MYMEMORY_URL || 'https://api.mymemory.translated.net/get';
    this.googleApiKey = process.env.GOOGLE_TRANSLATE_API_KEY || '';
    this.deeplApiKey = process.env.DEEPL_API_KEY || '';
  }

  isConfigured() {
    return this.enabled;
  }

  async translateToEnglish(text) {
    return this.translate(text, 'EN');
  }

  async translate(text, targetLanguage = 'EN') {
    if (!this.enabled || !text) return text;

    const cacheKey = `${targetLanguage}::${text}`;
    if (translationCache.has(cacheKey)) {
      return translationCache.get(cacheKey);
    }

    let result = text;

    // Try each service in fallback order
    result = await this.tryLibreTranslate(text, targetLanguage, result);
    if (result !== text) {
      translationCache.set(cacheKey, result);
      return result;
    }

    result = await this.tryMyMemory(text, targetLanguage, result);
    if (result !== text) {
      translationCache.set(cacheKey, result);
      return result;
    }

    result = await this.tryGoogleTranslate(text, targetLanguage, result);
    if (result !== text) {
      translationCache.set(cacheKey, result);
      return result;
    }

    result = await this.tryDeepL(text, targetLanguage, result);
    translationCache.set(cacheKey, result);
    return result;
  }

  async tryLibreTranslate(text, targetLanguage, fallback) {
    if (process.env.LIBRETRANSLATE_ENABLED === 'false') return fallback;
    try {
      const payload = {
        q: text,
        source: 'auto',
        target: targetLanguage.toLowerCase(),
      };
      const response = await axios.post(this.libreTranslateUrl, payload, { timeout: 30000 });
      if (response.data && response.data.translatedText) {
        console.log(`LibreTranslate: Translated ${text.length} chars`);
        return response.data.translatedText;
      }
    } catch (error) {
      console.debug(`LibreTranslate failed: ${error.message}`);
    }
    return fallback;
  }

  async tryMyMemory(text, targetLanguage, fallback) {
    if (process.env.MYMEMORY_ENABLED === 'false') return fallback;
    try {
      const sourceLang = detectLanguage(text);
      const targetLang = targetLanguage === 'EN' ? 'en' : targetLanguage.toLowerCase();
      
      // Skip if already in target language
      if (sourceLang === targetLang) return fallback;
      
      const url = `${this.myMemoryUrl}?q=${encodeURIComponent(text)}&langpair=${sourceLang}|${targetLang}`;
      const response = await axios.get(url, { timeout: 30000 });
      if (response.data && response.data.responseData && response.data.responseData.translatedText && response.data.responseData.translatedText !== text) {
        console.log(`MyMemory: Translated ${text.length} chars`);
        return response.data.responseData.translatedText;
      }
    } catch (error) {
      console.debug(`MyMemory failed: ${error.message}`);
    }
    return fallback;
  }

  async tryGoogleTranslate(text, targetLanguage, fallback) {
    if (process.env.GOOGLE_TRANSLATE_ENABLED !== 'true' || !this.googleApiKey || this.googleApiKey.startsWith('YOUR_')) {
      return fallback;
    }
    try {
      const payload = {
        q: text,
        target: targetLanguage.toLowerCase(),
        key: this.googleApiKey,
      };
      const response = await axios.post('https://translation.googleapis.com/language/translate/v2', payload, { timeout: 30000 });
      if (response.data && response.data.data && response.data.data.translations && response.data.data.translations.length > 0) {
        console.log(`Google Translate: Translated ${text.length} chars`);
        return response.data.data.translations[0].translatedText;
      }
    } catch (error) {
      console.debug(`Google Translate failed: ${error.message}`);
    }
    return fallback;
  }

  async tryDeepL(text, targetLanguage, fallback) {
    if (process.env.DEEPL_ENABLED !== 'true' || !this.deeplApiKey || this.deeplApiKey.startsWith('YOUR_')) {
      return fallback;
    }
    try {
      const response = await axios.post(
        process.env.DEEPL_URL || 'https://api-free.deepl.com/v1/translate',
        {
          text,
          target_lang: targetLanguage,
          auth_key: this.deeplApiKey,
        },
        {
          headers: {
            'Content-Type': 'application/json'
          },
          timeout: 30000
        }
      );
      if (response.data && response.data.translations && response.data.translations.length > 0) {
        console.log(`DeepL: Translated ${text.length} chars`);
        return response.data.translations[0].text;
      }
    } catch (error) {
      console.debug(`DeepL failed: ${error.message}`);
    }
    return fallback;
  }

  clearCache() {
    translationCache.clear();
  }
}

module.exports = TranslationService;
